### PR TITLE
Add migration docs about behat tests

### PR DESCRIPTION
**Pull request by zuk3975**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Description?  | Add migration docs about behavior tests
| Fixed ticket? |

### CODE REVIEW

File: src/content/1.7/development/architecture/migration-guide/testing/_index.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
+---
+title: Testing
+menuTitle: Testing
+---
+
+# Testing
+
+You are encouraged to add both unit and functional tests for every new class
```
Comment (matks) [inline src/content/1.7/development/architecture/migration-guide/testing/_index.md:8]: @Progi1984 will ❤️

**Review by matks — APPROVED** [review]
Looks good 👍

File: src/content/1.7/development/architecture/migration-guide/testing/_index.md (hunk 1 above)
Comment (zuk3975) [reply src/content/1.7/development/architecture/migration-guide/testing/_index.md:8]: haha, I just moved it from another file. (actually I think we are not adding smoke tests for every class :smile:)

File: src/content/1.7/development/architecture/migration-guide/testing/behat.md [hunk 2]
```diff
@@ -0,0 +1,176 @@
+---
+title: Behat testing
+menuTitle: Behat testing
+weight: 20
+---
+
+# Behat testing
+
+{{% notice tip %}}
+It is important that you get familiar with the `Behats` principles before reading further. You can find the official behat documentation [here](https://docs.behat.org/en/latest/guides.html).
```
Comment (Progi1984) [inline src/content/1.7/development/architecture/migration-guide/testing/behat.md:10]:
```suggestion
It is important that you get familiar with the `Behat` principles before reading further. You can find the official Behat documentation [here](https://docs.behat.org/en/latest/guides.html).
```

Comment (Progi1984) [inline src/content/1.7/development/architecture/migration-guide/testing/behat.md:10]:
```suggestion
It is important that you get familiar with the `Behats` principles before reading further. You can find the official Behat documentation [here](https://docs.behat.org/en/latest/guides.html).
```

File: src/content/1.7/development/architecture/migration-guide/testing/behat.md [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+{{% /notice %}}
+
+A behaviour (`behat`) tests are a part of integration tests. They allow testing how multiple components are working together. In PrestaShop, behats are used to test the `Application` and `Domain` layer integration - basically all the [CQRS](/1.7/development/architecture/domain/cqrs) commands and queries.
+
+## Database
+During behat tests the actual database queries are executed, therefore before testing you need to run a command `composer create-test-db` to create a test database.
+
+{{% notice %}}
+The `create-test-db` script installs a fresh prestashop with a fixtures in a new database called `test_{your database name}` and dumps the database in your machine `/tmp` directory named `ps_dump.sql`. That `ps_dump.sql` is later used to reset the database. You can check the actual script for more information - [/tests-legacy/create-test-db.php](https://github.com/PrestaShop/PrestaShop/blob/1.7.8.x/tests-legacy/create-test-db.php).
```
Comment (Progi1984) [inline src/content/1.7/development/architecture/migration-guide/testing/behat.md:19]:
```suggestion
The `create-test-db` script installs a fresh prestashop with fixtures in a new database called `test_{your database name}` and dumps the database in your machine `/tmp` directory named `ps_dump.sql`. That `ps_dump.sql` is later used to reset the database. You can check the actual script for more information - [/tests-legacy/create-test-db.php](https://github.com/PrestaShop/PrestaShop/blob/1.7.8.x/tests-legacy/create-test-db.php).
```

File: src/content/1.7/development/architecture/migration-guide/testing/behat.md [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+{{% /notice %}}
+
+## Behats structure in PrestaShop
+
+Behat related files are located in [tests/Integration/Behaviour/](https://github.com/PrestaShop/PrestaShop/tree/1.7.8.x/tests/Integration/Behaviour). This directory contains following files:
+- [behat.yml]({{< relref "#behatyml">}}) - this is the test suites configuration file which describes feature paths and contexts for every test suite. It can be passed as an argument when running the tests like this `./vendor/bin/behat -c tests/Integration/Behaviour/behat.yml`.
+- bootstrap.php - this file loads the `Kernel` for a behat tests environment.
+- Features - this directory contains all the [Scenarios](https://github.com/PrestaShop/PrestaShop/tree/1.7.8.x/tests/Integration/Behaviour/Features/Scenario) and [Contexts](https://github.com/PrestaShop/PrestaShop/tree/1.7.8.x/tests/Integration/Behaviour/Features/Context). More about it [bellow]({{< relref "#features" >}}).
```
Comment (Progi1984) [inline src/content/1.7/development/architecture/migration-guide/testing/behat.md:27]:
```suggestion
- Features - this directory contains all the [Scenarios](https://github.com/PrestaShop/PrestaShop/tree/1.7.8.x/tests/Integration/Behaviour/Features/Scenario) and [Contexts](https://github.com/PrestaShop/PrestaShop/tree/1.7.8.x/tests/Integration/Behaviour/Features/Context). More about it [below]({{< relref "#features" >}}).
```

File: src/content/1.7/development/architecture/migration-guide/testing/behat.md [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+
+## Features
+
+{{% notice tip %}}
+Before continuing, **please read the official `behat` documentation** about the [features and scenarios](https://docs.behat.org/en/latest/user_guide/features_scenarios.html).
+{{% /notice %}}
+
+In PrestaShop all `*.feature` files are placed in [.tests/Integration/Behaviour/Features/Scenario](https://github.com/PrestaShop/PrestaShop/tree/1.7.8.x/tests/Integration/Behaviour/Features/Scenario), each feature is placed in a dedicated directory organized by `domain` (or even a `subdomain` if necessary). These feature files contains text that describes the testing scenarios in a user-friendly manner, each of them must start with a keyword `Feature` and have a one or multiple scenarios starting with a keyword `Scenario`. For example:
```
Comment (Progi1984) [inline src/content/1.7/development/architecture/migration-guide/testing/behat.md:35]:
```suggestion
In PrestaShop, all `*.feature` files are placed in [.tests/Integration/Behaviour/Features/Scenario](https://github.com/PrestaShop/PrestaShop/tree/1.7.8.x/tests/Integration/Behaviour/Features/Scenario). Each feature is placed in a dedicated directory organized by `domain` (or even a `subdomain` if necessary). These feature files contains text that describes the testing scenarios in a user-friendly manner, each of them must start with a keyword `Feature` and have a one or multiple scenarios starting with a keyword `Scenario`. For example:
```

File: src/content/1.7/development/architecture/migration-guide/testing/behat.md [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+
+This is how the context looks like:
+```php
+class OrderFeatureContext extends AbstractDomainFeatureContext
+{
+//    ...
+    /**
+     * @Given I add order :orderReference with the following details:
+     *
+     * @param $orderReference
```
Comment (Progi1984) [inline src/content/1.7/development/architecture/migration-guide/testing/behat.md:81]:
```suggestion
     * @param string $orderReference
```

File: src/content/1.7/development/architecture/migration-guide/testing/behat.md [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+     * @param TableNode $table
+     */
+    public function addOrderWithTheFollowingDetails($orderReference, TableNode $table)
```
Comment (Progi1984) [inline src/content/1.7/development/architecture/migration-guide/testing/behat.md:84]:
```suggestion
    public function addOrderWithTheFollowingDetails(string $orderReference, TableNode $table): void
```

File: src/content/1.7/development/architecture/migration-guide/testing/behat.md [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+    
+//    ...
+
+```
+
+As you can see in example, the string `@Given I add order :orderReference with the following details:` maps this method to related line in `*.feature` file. The `:orderReference` acts as a variable which actually is the `id` of the order, that is saved into the [`SharedStorage`]({{< relref "#shared-storage" >}}). The `TableNode $table` is a specific argument, you can read about it [here](https://behat.org/en/latest/user_guide/writing_scenarios.html#tables).
+
+## Shared storage
+
+The [SharedStorage](https://github.com/PrestaShop/PrestaShop/blob/1.7.8.x/tests/Integration/Behaviour/Features/Context/SharedStorage.php) is responsible for holding certain values in memory which are shared across the feature. The most common usage example is the `id` reference - we specify a certain keyword e.g. `prouduct1` before creating it, and once the command returns the auto-incremented value, we set it in shared storage like this `SharedStorage::getStorage()->set($orderReference, $orderId->getValue());`. In upcoming scenarios we can reuse this reference to get the record, something like this:
```
Comment (Progi1984) [inline src/content/1.7/development/architecture/migration-guide/testing/behat.md:112]:
```suggestion
The [SharedStorage](https://github.com/PrestaShop/PrestaShop/blob/1.7.8.x/tests/Integration/Behaviour/Features/Context/SharedStorage.php) is responsible for holding certain values in memory which are shared across the feature. The most common usage example is the `id` reference - we specify a certain keyword e.g. `product1` before creating it, and once the command returns the auto-incremented value, we set it in shared storage like this `SharedStorage::getStorage()->set($orderReference, $orderId->getValue());`. In upcoming scenarios we can reuse this reference to get the record, something like this:
```

File: src/content/1.7/development/architecture/migration-guide/testing/behat.md [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+@clear-cache-before-feature
+@clear-cache-after-feature
+Feature: Add basic product from Back Office (BO)
+  As a BO user
+  I need to be able to add new product with basic information from the BO
+
+```
+
+{{% notice tip %}}
+You can also tag specific `features` if you want to run only them with a "--tags" filter. For example, if you add following tag in your Feature:
```
Comment (Progi1984) [inline src/content/1.7/development/architecture/migration-guide/testing/behat.md:139]:
```suggestion
You can also tag specific `features` if you want to run only them with a `--tags` filter. For example, if you add following tag in your Feature:
```

**Review by Progi1984 — COMMENTED** [review]
Some feedbacks

File: src/content/1.7/development/architecture/migration-guide/testing/behat.md (hunk 2 above)
Comment (zuk3975) [reply src/content/1.7/development/architecture/migration-guide/testing/behat.md:10]: I will make it lowercase letters everywhere.
Screenshot from 2021-05-13 17-03-15

Comment (zuk3975) [reply src/content/1.7/development/architecture/migration-guide/testing/behat.md:10]: I will make it lowercase letters everywhere and maybe wrap in backslashes or make it bold to be more visible.
Screenshot from 2021-05-13 17-03-15

**Review by Progi1984 — APPROVED** [review]

**Comment by Progi1984** [comment]
Thanks @zuk3975

**Pull request merged**